### ISSUE TITLE
A third episode option for accuracy

### ISSUE THREAD

**Issue by BlastBolt5** [crunchy-labs/crunchy-cli#246]
**Is your feature request related to a problem? Please describe.**
As was stated by me in #206 and ohitstom in #241, having the numbers grow with {episode_number} and only having relative numbers with {relative_episode_number} has major issues with media servers like plex or jellyfin, because they don't check for numbers relative to the show or the season, so a season 2 with an episode 0 or a .5 episode is gonna be named entirely wrong to a point where I would have to go and rename 12+ episodes a season.

**Describe the solution you'd like**
The first solution that comes to mind would just be a accurate relative episode option that would set the episode number to exactly what it would be in the seasons. another option would be to be have a setting to ignore 0 and .5 episodes.

**Additional context**
For right now i'm probably gonna right a little script that just takes the episodes and subtracts it from the total of episodes from the previous seasons minus any 0 or .5 episodes. i understand this is an issue with how crunchy roll numbers their episodes (some with resets between seasons and some not) but rather than requiring a separate script it would be amazing to have this baked right in.

**Comment by alazare619** [comment]
I'm going to expand on an option. Currently right now this feature request is not possible. Crunchy-cli only pulls the data from the crunchy website and alot of shows are blatantly wrong. a Better option is to integrate the anidb api. AniDB currently has every show indexed on their site. With 1 query of show name it will list all the episodes and where to view them it then just needs to match the url to the crunchyroll URL and it has scene information necessary to tag a episode 100% correct in the way that sonarr/plex/jellyfin etc all expect. This is the PROPER solution and nothing else will really suffice. It is soley the fault of crunchyroll for posting the content incorrectly.

**Comment by alazare619** [comment]
@ByteDream

**Comment by BlastBolt5** [comment]
an option to have relative episode number with out episode 0s or .5s would also be a fix to the main issue, but using that anidb connection would also be amazing.

**Comment by ByteDream** [comment]
A AniDB integration would be indeed nice.
As far as I can see, anidb has a field in one of their api endpoints which actually resolves to Crunchyroll episode ids.
![image](https://github.com/crunchy-labs/crunchy-cli/assets/63594396/f3a1773b-8955-4a50-b9e0-f4a4305dc250)
I assume that `type="28"` means that the resource "provider" is crunchyroll and the id in `identifier` tag matches the internal crunchyroll id of the (japanese) first episode of darling in the franxx.

I'm only concerned about the endpoints anidb provides to access the api. They publicly provide a SOAP or raw UDP endpoint which both are not ideal for modern development (imo).

**Comment by ByteDream** [comment]
> A AniDB integration _would_ be indeed nice.

After a bit of searching multiple problems occurred. It is not guaranteed that the Crunchyroll title is the same as the title in AniDB. For example if you're located in Germany, only the german dub of sword art online is available and every season of it has a `(DE)` prefix or `(German Dub)` suffix (depending if your display language is set to german or english). The AniDB entries doesn't have this prefix/suffix.
Sure, crunchy-cli can try to remove any suspicious prefix/suffix, but I'm sure that this will lead to other problems where something gets removed which actually belongs to the correct title.
Another thing are special characters. In some cases titles have special characters like apostrophes. On Crunchyroll the title has `'` as apostrophe, whereas AniDB has <code>`</code>. And here it's the same problem as described above, sure the character can be removed from the title but this may brake other things.

**Comment by ByteDream** [comment]
![image](https://github.com/crunchy-labs/crunchy-cli/assets/63594396/965e3439-29fb-42e7-8157-80619306f41b)
![image](https://github.com/crunchy-labs/crunchy-cli/assets/63594396/10d199af-9701-443d-b40c-822a2a2909d9)

Okay I have a question. For example, the first Re:Zero episode has `E1A` in it's title and the sequence number `0.5`. With a option which handles such cases, should it return `E1A` or `E0.5`?

**Comment by BlastBolt5** [comment]
Well in this case I think it should just be dropped, Episode 1 here is a single episode that includes both episode 1a and 1b, But overall I think cause how letters are used by crunchyroll they should come out as be spit out as 1a and 1b

**Comment by ByteDream** [comment]
I've added a option which should handle the most cases. You can test it by cloning the repo and checking out the `feature/relative_sequence_number` branch or using one of the [pre-built binaries](https://github.com/crunchy-labs/crunchy-cli/actions/runs/6525878928).
The new format option is called `{relative_sequence_number}` as it is based of the `sequence_number` api field but it might get renamed when I merge it, to state more clearly what it does.
It would be great if you guys could test this and give feedback @BlastBolt5 @alazare619 @ohitstom

**Comment by ohitstom** [comment]
> I've added a option which should handle the most cases. You can test it by cloning the repo and checking out the `feature/relative_sequence_number` branch or using one of the [pre-built binaries](https://github.com/crunchy-labs/crunchy-cli/actions/runs/6525878928).
> The new format option is called `{relative_sequence_number}` as it is based of the `sequence_number` api field but it might get renamed when I merge it, to state more clearly what it does.
> It would be great if you guys could test this and give feedback @BlastBolt5 @alazare619 @ohitstom

I'll try and find some time to test it sounds awesome tysm!

**Comment by BlastBolt5** [comment]
Still doing some testing but this is what i'm seeing so far:

1. Forward slashes in in the formatting for -o do not appear for outputted file (can't sort into folders)
2. First episodes of a season with a .5 are formatted to E0 but later season .5 episodes overwrite the episode that should be downloaded, like E24.5 is downloaded as E24 and then the actual E24 just isn't downloaded
3. {relative_sequence_number} is not formatted to have 2 digits by default
4. Idk if this is related to this but when doing [S1E24-] for [reincarnated as a slime](https://www.crunchyroll.com/series/GYZJ43JMR/that-time-i-got-reincarnated-as-a-slime) episode 24.5 doesn't show up to be downloaded

**Comment by ohitstom** [comment]
i think it would be nice if this option named anything that is a float as a special instead, that would solve 100% of the issue on my end since most of the time these episodes are classed as specials regardless

**Comment by BlastBolt5** [comment]
That also would probably be the best solution for me, and another option for 1a and 1b episodes would be outputting them as e1p1 and e1p2 possibly

**Comment by ByteDream** [comment]
> Still doing some testing but this is what i'm seeing so far:
> 
>   1. Forward slashes in in the formatting for -o do not appear for outputted file (can't sort into folders)
> 
>   2. First episodes of a season with a .5 are formatted to E0 but later season .5 episodes overwrite the episode that should be downloaded, like E24.5 is downloaded as E24 and then the actual E24 just isn't downloaded
> 
>   3. {relative_sequence_number} is not formatted to have 2 digits by default
> 
>   4. Idk if this is related to this but when doing [S1E24-] for [reincarnated as a slime](https://www.crunchyroll.com/series/GYZJ43JMR/that-time-i-got-reincarnated-as-a-slime) episode 24.5 doesn't show up to be downloaded
> 
> 
> Idk which parts of that is related to this specific branch but just thought i'd mention the issues

I think everything of the named issues are resolved now. Could you verify my changes?

**Comment by ByteDream** [comment]
> i think it would be nice if this option named anything that is a float as a special instead, that would solve 100% of the issue on my end since most of the time these episodes are classed as specials regardless, same goes for "1a" type episodes, if it returns sequence_number "anything.anything" make it a special

How should an episode be marked as special?

**Comment by BlastBolt5** [comment]
> > i think it would be nice if this option named anything that is a float as a special instead, that would solve 100% of the issue on my end since most of the time these episodes are classed as specials regardless, same goes for "1a" type episodes, if it returns sequence_number "anything.anything" make it a special
> 
> How should an episode be marked as special?

The best way would be if the season for that episode could be set to special or 00

**Comment by BlastBolt5** [comment]
> > Still doing some testing but this is what i'm seeing so far:
> > 
> > 1. Forward slashes in in the formatting for -o do not appear for outputted file (can't sort into folders)
> > 2. First episodes of a season with a .5 are formatted to E0 but later season .5 episodes overwrite the episode that should be downloaded, like E24.5 is downloaded as E24 and then the actual E24 just isn't downloaded
> > 3. {relative_sequence_number} is not formatted to have 2 digits by default
> > 4. Idk if this is related to this but when doing [S1E24-] for [reincarnated as a slime](https://www.crunchyroll.com/series/GYZJ43JMR/that-time-i-got-reincarnated-as-a-slime) episode 24.5 doesn't show up to be downloaded
> > 
> > Idk which parts of that is related to this specific branch but just thought i'd mention the issues
> 
> I think everything of the named issues are resolved now. Could you verify my changes? (over at #257, that's the active PR)

Yep just tested the new commit and all 4 of those issues are fixed, tysm!

**Comment by ohitstom** [comment]
> > i think it would be nice if this option named anything that is a float as a special instead, that would solve 100% of the issue on my end since most of the time these episodes are classed as specials regardless, same goes for "1a" type episodes, if it returns sequence_number "anything.anything" make it a special
> 
> How should an episode be marked as special?

if it is a float number, set its sequence number to just the word "special", since lots of metadata providers dont follow half episodes are more the common "special - title" format, such as adventure time (not applicable of course but just off the top of my head)

though this would likely fit better in its own flag that could work in tandem with relative sequence, maybe something named rename_floats or something

**Comment by ByteDream** [comment]
Okay I see. I don't like the idea of an extra flag for this but I can't see another way atm.

To give a good amount of flexibility, the extra flag could be designed exactly like `-o` with the only difference that it only gets used when the sequence number is a float.

E.g.(assuming that the new flag is called `-os`):
```shell
$ crunchy-cli archive -o "S{season_number}E{relative_sequence_number} - {title}.mkv" -os "Special - {title}" ...
# S1E1 - ...
# S1E2 - ...
# Special - ...
# S1E3 - ...
```

How do you thing about this?

**Comment by BlastBolt5** [comment]
That looks awesome thank you so much

**Comment by BlastBolt5** [comment]
would there be anyway to make a flag to just ignore those episodes as well? Like how skip existing has one?

**Comment by ByteDream** [comment]
I've pushed new code to the PR which adds the `--special-output` flag to control the output file name of special episodes as well as a flag to ignore special episodes entirely (what @BlastBolt5 suggested)

**Comment by ByteDream** [comment]
Support for this landed in [v3.1.0](https://github.com/crunchy-labs/crunchy-cli/releases/tag/v3.1.0)

**Comment by BlastBolt5** [comment]
Thank you so much, I have been busy with school so I haven't been able to test this yet but it looks perfect
